File: packages/typescript/src/compliance/ccpa.ts
import { checkLicense } from "../rag/license-check";
/**
 * CCPA/CPRA Compliance Module (California Consumer Privacy Act)
 *
 * DISCLAIMER: This module provides consumer request handling utilities.
 * It does NOT replace legal review of your privacy policy or data practices.
 * Consult legal counsel for CCPA/CPRA compliance requirements.
 *
 * Implements:
 * - Right to know (data access)
 * - Right to delete
 * - Right to opt-out of sale/sharing
 * - Right to correct
 * - Automated decision-making opt-out (ADMT)
 * - Sensitive data consent tracking
 * - "Do Not Sell or Share" signal handling
 */

import type { Database } from "../database";

export interface CCPAConfig {
  enabled: boolean;
  /** Honor Global Privacy Control (GPC) opt-out signals */
  honorGPC?: boolean;
  /** Track "Do Not Sell or Share" preferences */
  trackOptOut?: boolean;
  /** Log automated decision-making for ADMT compliance */
  logADMT?: boolean;
}

export interface ConsumerRequest {
  type: "access" | "delete" | "correct" | "opt-out" | "opt-in";
  consumerId: string;
  requestedAt: string;
  completedAt?: string;
  status: "pending" | "verified" | "completed" | "denied";
}

export class CCPAManager {
  private db: Database;
  private config: CCPAConfig;

  constructor(db: Database, config: CCPAConfig) {
    checkLicense("Compliance"); this.db = db;
    this.config = config;

    try {
      db.run(`CREATE TABLE IF NOT EXISTS bulwark_ccpa_requests (
        id TEXT PRIMARY KEY,
        type TEXT NOT NULL,
        consumer_id TEXT NOT NULL,
        status TEXT NOT NULL DEFAULT 'pending',
        data TEXT,
        requested_at TEXT NOT NULL DEFAULT (datetime('now')),
        completed_at TEXT
      )`);
      db.run("CREATE INDEX IF NOT EXISTS idx_ccpa_consumer ON bulwark_ccpa_requests(consumer_id)");

      db.run(`CREATE TABLE IF NOT EXISTS bulwark_ccpa_optout (
        consumer_id TEXT PRIMARY KEY,
        opted_out INTEGER NOT NULL DEFAULT 0,
        opt_out_date TEXT,
        gpc_signal INTEGER NOT NULL DEFAULT 0,
        updated_at TEXT NOT NULL DEFAULT (datetime('now'))
      )`);
    } catch {}
  }

  /** Right to Know — return all data for a consumer */
  accessRequest(consumerId: string): { data: unknown[]; requestId: string } {
    const id = `ccpa_${Date.now()}`;
    const auditData = this.db.queryAll("SELECT * FROM bulwark_audit WHERE user_id = ? ORDER BY timestamp DESC", [consumerId]);
    const usageData = this.db.queryAll("SELECT * FROM bulwark_usage WHERE user_id = ? ORDER BY timestamp DESC", [consumerId]);

    this.db.run("INSERT INTO bulwark_ccpa_requests (id, type, consumer_id, status, completed_at) VALUES (?, 'access', ?, 'completed', datetime('now'))", [id, consumerId]);

    return { data: [...auditData, ...usageData], requestId: id };
  }

  /** Right to Delete — erase all consumer data */
  deleteRequest(consumerId: string): { requestId: string; deleted: boolean } {
    const id = `ccpa_${Date.now()}`;
    this.db.run("DELETE FROM bulwark_audit WHERE user_id = ?", [consumerId]);
    this.db.run("DELETE FROM bulwark_usage WHERE user_id = ?", [consumerId]);

    this.db.run("INSERT INTO bulwark_ccpa_requests (id, type, consumer_id, status, completed_at) VALUES (?, 'delete', ?, 'completed', datetime('now'))", [id, consumerId]);

    return { requestId: id, deleted: true };
  }

  /** Right to Opt-Out of Sale/Sharing */
  optOut(consumerId: string, gpcSignal = false): void {
    this.db.run(
      "INSERT OR REPLACE INTO bulwark_ccpa_optout (consumer_id, opted_out, opt_out_date, gpc_signal, updated_at) VALUES (?, 1, datetime('now'), ?, datetime('now'))",
      [consumerId, gpcSignal ? 1 : 0]
    );
  }

  /** Check if consumer has opted out */
  isOptedOut(consumerId: string): boolean {
    const row = this.db.queryOne<{ opted_out: number }>(
      "SELECT opted_out FROM bulwark_ccpa_optout WHERE consumer_id = ?", [consumerId]
    );
    return !!row?.opted_out;
  }

  /** Handle Global Privacy Control signal */
  handleGPC(consumerId: string): void {
    if (!this.config.honorGPC) return;
    this.optOut(consumerId, true);
  }

  /** Get all consumer requests (for compliance reporting) */
  getRequests(consumerId?: string, limit = 100): ConsumerRequest[] {
    const sql = consumerId
      ? "SELECT * FROM bulwark_ccpa_requests WHERE consumer_id = ? ORDER BY requested_at DESC LIMIT ?"
      : "SELECT * FROM bulwark_ccpa_requests ORDER BY requested_at DESC LIMIT ?";
    return this.db.queryAll(sql, consumerId ? [consumerId, limit] : [limit]);
  }

  /** Generate CCPA compliance report */
  generateReport(): {
    totalRequests: number;
    byType: Record<string, number>;
    optedOutConsumers: number;
    gpcSignals: number;
    avgResponseTime: string;
  } {
    const total = this.db.queryOne<{ c: number }>("SELECT COUNT(*) as c FROM bulwark_ccpa_requests");
    const byType = this.db.queryAll<{ type: string; c: number }>("SELECT type, COUNT(*) as c FROM bulwark_ccpa_requests GROUP BY type");
    const optedOut = this.db.queryOne<{ c: number }>("SELECT COUNT(*) as c FROM bulwark_ccpa_optout WHERE opted_out = 1");
    const gpc = this.db.queryOne<{ c: number }>("SELECT COUNT(*) as c FROM bulwark_ccpa_optout WHERE gpc_signal = 1");

    return {
      totalRequests: total?.c || 0,
      byType: Object.fromEntries(byType.map(r => [r.type, r.c])),
      optedOutConsumers: optedOut?.c || 0,
      gpcSignals: gpc?.c || 0,
      avgResponseTime: "< 45 days", // CCPA requires response within 45 days
    };
  }
}

File: packages/python/bulwark_ai/compliance/ccpa.py
"""
CCPA/CPRA Compliance Module (California Consumer Privacy Act)

DISCLAIMER: This module provides consumer request handling utilities.
It does NOT replace legal review of your privacy policy or data practices.
Consult legal counsel for CCPA/CPRA compliance requirements.

Implements:
- Right to know (data access)
- Right to delete
- Right to opt-out of sale/sharing
- Global Privacy Control (GPC) signal handling
- Consumer request tracking
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

import aiosqlite


@dataclass
class CCPAConfig:
    enabled: bool = True
    honor_gpc: bool = True
    track_opt_out: bool = True
    log_admt: bool = False


@dataclass
class ConsumerRequest:
    type: str  # "access" | "delete" | "correct" | "opt-out" | "opt-in"
    consumer_id: str
    requested_at: str
    completed_at: Optional[str] = None
    status: str = "pending"  # "pending" | "verified" | "completed" | "denied"


class CCPAManager:
    """
    CCPA/CPRA compliance utilities.

    DISCLAIMER: This module provides consumer request handling utilities.
    It does NOT replace legal review of your privacy policy or data practices.
    Consult legal counsel for CCPA/CPRA compliance requirements.
    """

    def __init__(self, db_path: str, config: CCPAConfig | None = None) -> None:
        self._db_path = db_path
        self._config = config or CCPAConfig()

    async def initialize(self) -> None:
        """Create required tables."""
        async with aiosqlite.connect(self._db_path) as db:
            await db.execute("""
                CREATE TABLE IF NOT EXISTS bulwark_ccpa_requests (
                    id TEXT PRIMARY KEY,
                    type TEXT NOT NULL,
                    consumer_id TEXT NOT NULL,
                    status TEXT NOT NULL DEFAULT 'pending',
                    data TEXT,
                    requested_at TEXT NOT NULL DEFAULT (datetime('now')),
                    completed_at TEXT
                )
            """)
            await db.execute(
                "CREATE INDEX IF NOT EXISTS idx_ccpa_consumer "
                "ON bulwark_ccpa_requests(consumer_id)"
            )
            await db.execute("""
                CREATE TABLE IF NOT EXISTS bulwark_ccpa_optout (
                    consumer_id TEXT PRIMARY KEY,
                    opted_out INTEGER NOT NULL DEFAULT 0,
                    opt_out_date TEXT,
                    gpc_signal INTEGER NOT NULL DEFAULT 0,
                    updated_at TEXT NOT NULL DEFAULT (datetime('now'))
                )
            """)
            await db.commit()

    async def _connect(self) -> aiosqlite.Connection:
        db = await aiosqlite.connect(self._db_path)
        db.row_factory = aiosqlite.Row
        return db

    async def access_request(self, consumer_id: str) -> Dict[str, Any]:
        """Right to Know -- return all data for a consumer."""
        req_id = f"ccpa_{int(time.time() * 1000)}"

        async with await self._connect() as db:
            # Gather consumer data
            audit_data = await self._query_all(
                db,
                "SELECT * FROM bulwark_audit WHERE user_id = ? ORDER BY timestamp DESC",
                (consumer_id,),
            )
            usage_data = await self._query_all(
                db,
                "SELECT * FROM bulwark_usage WHERE user_id = ? ORDER BY timestamp DESC",
                (consumer_id,),
            )

            # Log the request
            await db.execute(
                "INSERT INTO bulwark_ccpa_requests (id, type, consumer_id, status, completed_at) "
                "VALUES (?, 'access', ?, 'completed', datetime('now'))",
                (req_id, consumer_id),
            )
            await db.commit()

        return {"data": audit_data + usage_data, "request_id": req_id}

    async def delete_request(self, consumer_id: str) -> Dict[str, Any]:
        """Right to Delete -- erase all consumer data."""
        req_id = f"ccpa_{int(time.time() * 1000)}"

        async with await self._connect() as db:
            await db.execute("DELETE FROM bulwark_audit WHERE user_id = ?", (consumer_id,))
            await db.execute("DELETE FROM bulwark_usage WHERE user_id = ?", (consumer_id,))
            await db.execute(
                "INSERT INTO bulwark_ccpa_requests (id, type, consumer_id, status, completed_at) "
                "VALUES (?, 'delete', ?, 'completed', datetime('now'))",
                (req_id, consumer_id),
            )
            await db.commit()

        return {"request_id": req_id, "deleted": True}

    async def opt_out(self, consumer_id: str, gpc_signal: bool = False) -> None:
        """Right to Opt-Out of Sale/Sharing."""
        async with await self._connect() as db:
            await db.execute(
                "INSERT OR REPLACE INTO bulwark_ccpa_optout "
                "(consumer_id, opted_out, opt_out_date, gpc_signal, updated_at) "
                "VALUES (?, 1, datetime('now'), ?, datetime('now'))",
                (consumer_id, 1 if gpc_signal else 0),
            )
            await db.commit()

    async def is_opted_out(self, consumer_id: str) -> bool:
        """Check if consumer has opted out."""
        async with await self._connect() as db:
            async with db.execute(
                "SELECT opted_out FROM bulwark_ccpa_optout WHERE consumer_id = ?",
                (consumer_id,),
            ) as cur:
                row = await cur.fetchone()
        return bool(row and row[0])

    async def handle_gpc(self, consumer_id: str) -> None:
        """Handle Global Privacy Control signal."""
        if not self._config.honor_gpc:
            return
        await self.opt_out(consumer_id, gpc_signal=True)

    async def get_requests(
        self, consumer_id: str | None = None, limit: int = 100
    ) -> List[ConsumerRequest]:
        """Get all consumer requests (for compliance reporting)."""
        async with await self._connect() as db:
            if consumer_id:
                sql = "SELECT * FROM bulwark_ccpa_requests WHERE consumer_id = ? ORDER BY requested_at DESC LIMIT ?"
                params: tuple[Any, ...] = (consumer_id, limit)
            else:
                sql = "SELECT * FROM bulwark_ccpa_requests ORDER BY requested_at DESC LIMIT ?"
                params = (limit,)

            async with db.execute(sql, params) as cur:
                cols = [d[0] for d in cur.description] if cur.description else []
                rows = await cur.fetchall()

        return [
            ConsumerRequest(
                type=dict(zip(cols, row))["type"],
                consumer_id=dict(zip(cols, row))["consumer_id"],
                requested_at=dict(zip(cols, row)).get("requested_at", ""),
                completed_at=dict(zip(cols, row)).get("completed_at"),
                status=dict(zip(cols, row)).get("status", "pending"),
            )
            for row in rows
        ]

    async def generate_report(self) -> Dict[str, Any]:
        """Generate CCPA compliance report."""
        async with await self._connect() as db:
            async with db.execute("SELECT COUNT(*) FROM bulwark_ccpa_requests") as cur:
                total_row = await cur.fetchone()
            total = total_row[0] if total_row else 0

            async with db.execute(
                "SELECT type, COUNT(*) as c FROM bulwark_ccpa_requests GROUP BY type"
            ) as cur:
                type_rows = await cur.fetchall()
            by_type = {row[0]: row[1] for row in type_rows}

            async with db.execute(
                "SELECT COUNT(*) FROM bulwark_ccpa_optout WHERE opted_out = 1"
            ) as cur:
                opted_row = await cur.fetchone()
            opted_out = opted_row[0] if opted_row else 0

            async with db.execute(
                "SELECT COUNT(*) FROM bulwark_ccpa_optout WHERE gpc_signal = 1"
            ) as cur:
                gpc_row = await cur.fetchone()
            gpc = gpc_row[0] if gpc_row else 0

        return {
            "total_requests": total,
            "by_type": by_type,
            "opted_out_consumers": opted_out,
            "gpc_signals": gpc,
            "avg_response_time": "< 45 days",  # CCPA requires response within 45 days
        }

    async def _query_all(
        self, db: aiosqlite.Connection, sql: str, params: tuple[Any, ...] = ()
    ) -> List[Dict[str, Any]]:
        try:
            async with db.execute(sql, params) as cur:
                cols = [d[0] for d in cur.description] if cur.description else []
                rows = await cur.fetchall()
            return [dict(zip(cols, row)) for row in rows]
        except Exception:
            return []
